ClassLikeCognitiveComplexityRule: take `final` and `extends` into account

clxmstaab opened the pull request:
closes https://github.com/symplify/symplify/issues/3425

clxmstaab commented:
@TomasVotruba feature-wise this is good to go.

we need to bikesheed a bit arround the question:

> This behavior should me IMO optional, as there made up rules at the moment. Something like "advanced mode", you know :)

I am not sure whether its usefull to create a separate mode for this feature. for my personal preference class-extending and non-final classes come with a cost and thats what we score here.

the numbers itself (so the actual calculated score) is artifial - or was the previous algorithm based on some kind of white-paper or well-known practice?

the score only means lower is better.. whether someone targets for e.g. 5 or 50 is personal preference.
wdyt?

clxmstaab commented:
while implementing I got the idea of also adding +10 for-each trait used on the class.. could also be a separate PR though

TomasVotruba commented on packages/phpstan-rules/packages/cognitive-complexity/src/AstCognitiveComplexityAnalyzer.php:
```diff
@@ -30,6 +33,16 @@ public function analyzeClassLike(ClassLike $classLike): int
+        if ($classLike->extends !== null) {
+            $totalCognitiveComplexity += self::INHERITANCE_CLASS_SCORE;
+        }
+
```
This should be located in separate service, to keep original complexity analyzer clean as possible. 

The extra service should not mix here, so we can easily maintain it and see original analyzer and the "extra" one.

TomasVotruba commented:
> or was the previous algorithm based on some kind of white-paper or well-known practice and we violate this rules now?

Yes, on 2 papers, the first I forgot and the second is in Sonarcube:
https://www.sonarsource.com/resources/white-papers/cognitive-complexity/

It should be linked in the docblock above the class, could you add it?
I probably removed it while refactoring from tokens to AST.

TomasVotruba commented:
> I am not sure whether its usefull to create a separate mode for this feature. for my personal preference class-extending and non-final classes come with a cost and thats what we score here.

This is extension over composition principle. The more we couple code together, the easier it gets to legacy. It might be easy to put all the code in one place, but it's hell for us maintainers :) 

Also "extension over composition" PRs are more-likely to be merged and are merged faster, because they're easier and more fun to read.

clxmstaab commented:
> It should be linked in the docblock above the class, could you add it?

added it.

> This is extension over composition principle. The more we couple code together, the easier it gets to legacy. It might be easy to put all the code in one place, but it's hell for us maintainers :)

just added a `CompositionOverInheritanceAnalyzer` and a separate Test-Suite for it

I think this is now good to go.

clxmstaab commented on packages/phpstan-rules/packages/cognitive-complexity/src/Rules/ClassLikeCognitiveComplexityRule.php:
```diff
@@ -38,9 +39,11 @@ final class ClassLikeCognitiveComplexityRule extends AbstractSymplifyRule implem
         private int $maxClassCognitiveComplexity = 50,
-        array $limitsByTypes = []
+        array $limitsByTypes = [],
+        private bool $scoreCompositionOverInheritance = false,
```
with this new paramter one can activate the composition-over-inheritance based analysis.

I left it OFF-by-default for BC

clxmstaab commented on packages/phpstan-rules/packages/cognitive-complexity/tests/AstCognitiveComplexityAnalyzer/AstCognitiveComplexityAnalyzerTest.php:
```diff
@@ -42,7 +42,7 @@ public function test(SmartFileInfo $fixtureFileInfo): void
         $inputAndExpected = StaticFixtureSplitter::splitFileInfoToInputAndExpected($fixtureFileInfo);
 
-        $functionLike = $this->parseFileToFistFunctionLike($inputAndExpected->getInput());
+        $functionLike = $this->parseFileToFirstFunctionLike($inputAndExpected->getInput());
```
fixed typo in method name: `Fist` -> `First`

TomasVotruba replied:
Thanks :+1:

clxmstaab commented on packages/phpstan-rules/packages/cognitive-complexity/tests/CompositionOverInheritanceAnalyzer/CompositionOverInheritanceAnalyzerTest.php:
```diff
@@ -0,0 +1,83 @@
+use Symplify\SmartFileSystem\SmartFileInfo;
+use Symplify\SymplifyKernel\Exception\ShouldNotHappenException;
+
+final class CompositionOverInheritanceAnalyzerTest extends TestCase
```
dedicated test-case, inspired by `AstCognitiveComplexityAnalyzerTest`

TomasVotruba replied:
Nice one :+1:

clxmstaab commented:
the last commit fixes the following phpstan errors:

![image](https://user-images.githubusercontent.com/47448731/127118099-ef2af379-a383-49c2-9531-26256e2fd86c.png)

not sure why these popup with this PR and did not before. I did not change the code related to this errors.

however, fixed it now.

clxmstaab commented on packages/phpstan-rules/packages/cognitive-complexity/tests/AstCognitiveComplexityAnalyzer/AstCognitiveComplexityAnalyzerTest.php:
```diff
@@ -65,16 +62,22 @@ public function provideTokensAndExpectedCognitiveComplexity(): Iterator
             fn (Node $node): bool => $node instanceof ClassMethod || $node instanceof Function_
         );
+
+        if (!$firstFunctionlike instanceof ClassMethod && !$firstFunctionlike instanceof Function_) {
```
assert the return-type early, so we don't need to handle a possible null-value at the call-site

TomasVotruba commented on packages/phpstan-rules/packages/cognitive-complexity/src/CompositionOverInheritanceAnalyzer.php:
```diff
@@ -0,0 +1,68 @@
+/**
+ * @see Symplify\PHPStanRules\CognitiveComplexity\Tests\CompositionOverInheritanceAnalyzer\CompositionOverInheritanceAnalyzerTest;
+ */
+final class CompositionOverInheritanceAnalyzer {
```
:+1: :clap:

TomasVotruba commented on packages/phpstan-rules/packages/cognitive-complexity/src/CompositionOverInheritanceAnalyzer.php:
```diff
@@ -0,0 +1,68 @@
+
+    public function analyzeClassLike(ClassLike $classLike): int
+    {
+        $totalCognitiveComplexity = 0;
+
+        $totalCognitiveComplexity += $this->astCognitiveComplexityAnalyzer->analyzeClassLike($classLike);
```
```suggestion
        $totalCognitiveComplexity = $this->astCognitiveComplexityAnalyzer->analyzeClassLike($classLike);
```

TomasVotruba commented on packages/phpstan-rules/packages/cognitive-complexity/tests/Rules/ClassLikeCognitiveComplexityRule/ClassLikeCognitiveComplexityRuleTest.php:
```diff
@@ -26,11 +26,13 @@ public function test(string $filePath, array $expectedErrorMessagesWithLines): v
         yield [__DIR__ . '/Fixture/SimpleCommand.php', [[$errorMessage, 9]]];
+
+        $errorMessage = sprintf(ClassLikeCognitiveComplexityRule::ERROR_MESSAGE, 'Class', 44, 5);
+        yield [__DIR__ . '/Fixture/NonFinalClass.php', [[$errorMessage, 9]]];
```
Could you separate this to a new test case file, so we have the original behavior in own file?

TomasVotruba commented:
I'm very happy with your refactoring :clap: 

The test case is the last thing to improve and we're ready to go

clxmstaab commented:
> Could you separate this to a new test case file, so we have the original behavior in own file?

just committed a separate `ClassLikeCompositionOverInheritanceRuleTest` which uses a separate config

TomasVotruba commented:
Thank you :clap:
